File: leetcode/[152] Maximum Product Subarray/dp.cpp
/*
 * @lc app=leetcode id=152 lang=cpp
 *
 * [152] Maximum Product Subarray
 */

// @lc code=start
/*** [vim-leetcode] For Local Syntax Checking ***/
#define DEPENDENCIES
#ifdef DEPENDENCIES
#include <iostream>
#include <iomanip>
#include <istream>
#include <ostream>
#include <sstream>
#include <stdio.h>
#include <vector>
#include <stack>
#include <map>
#include <unordered_map>
#include <set>
#include <unordered_set>
#include <list>
#include <forward_list>
#include <array>
#include <deque>
#include <queue>
#include <bitset>
#include <utility>
#include <algorithm>
#include <string>
#include <limits>
using namespace std;
#endif

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        //dp[i].first: the largest product which has nums[i] and possibly also previous item(s) as factor(s)
        //dp[i].second: the smallest product which has nums[i] and possibly also previous item(s) as factor(s)
        vector<pair<int, int>> dp(nums.size());
        int output;

        dp[0] = {nums[0], nums[0]};
        output = nums[0];
        for (int i = 1; i < nums.size(); i++){
            dp[i].first = max(max(nums[i], nums[i]*dp[i-1].first), nums[i]*dp[i-1].second);
            dp[i].second = min(min(nums[i], nums[i]*dp[i-1].first), nums[i]*dp[i-1].second);
            output = max(output, dp[i].first);
        }

        return output;
    }
};
// @lc code=end

File: leetcode/[1567] Maximum Length of Subarray With Positive Product/DP.java
/*
 * @lc app=leetcode id=1567 lang=java
 *
 * [1567] Maximum Length of Subarray With Positive Product
 */

// @lc code=start
/*** [vim-leetcode] For Local Syntax Checking ***/
import java.util.*;
import java.util.stream.*;
import java.util.Map.Entry;
import java.lang.*;

class Solution {
    public int getMaxLen(int[] nums) {
        int maxLen = 0;
        int negative = -1, positive = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {
                negative = -1;
                positive = 0;
            } else if (nums[i] > 0) {
                ++positive;
                negative += negative >= 0 ? 1 : 0;
            } else {
                int originalPositive = positive;
                positive = negative + 1;
                negative = originalPositive + 1;
            }
            maxLen = Math.max(positive, maxLen);
        }
        return maxLen;
    }
}
// @lc code=end
